### ISSUE TITLE
networkd WireGuard netdev: re-resolve DNS peer names every now and then (based on DNS TTL)

### ISSUE THREAD

**Issue by WGH-** [systemd/systemd#9911]
Currently, when endpoint of your Wireguard tunnel has both A and AAAA records, systemd-networkd will stick to one once, and will not switch to IPv4 once IPv6 becomes unavailable (and vice versa).

This can be considered a part of a bigger problem: it's probably not a good idea to use IP address resolved once forever regardless of DNS record TTL.

I think systemd-networkd should periodically (based on TTL) re-resolve Wireguard peer's IP, and also every time network configuration changes.

This problem has been briefly discussed in WireGuard mailing list with no clear consensus (whether it should be kernel's responsibility or userspace one) https://lists.zx2c4.com/pipermail/wireguard/2018-March/002467.html

**Comment by floppym** [comment]
I think input from @zx2c4 would be needed before changing anything here.

**Comment by zx2c4** [comment]
> I don't think I've seen any other software do this. 

OpenVPN has an option for it, actually. I think in the context of a VPN, re-resolving DNS actually makes a decent amount of sense for people connecting up to endpoints that have dynamic IPs when they're not persistently connected. The contrib directory ships with a [basic shell script for doing this](https://git.zx2c4.com/WireGuard/tree/contrib/examples/reresolve-dns).

**Comment by gdamjan** [comment]
my VPN server (a home server) has a dynamic ip address, changes once a day, and then wireguard is stuck. this would be a highly regarded feature.

I might take a look at the networkd code, but in the mean time does anyone have any pointers what and where to look into?

**Comment by el0hh** [comment]
Maybe the wireguard option "persistent keepalive" will help, to resolve the problem?

**Comment by grawity** [comment]
@gdamjan, just out of curiosity, what ISP forces a daily address change on *active* connections? Is that a common thing to do?

**Comment by JensErat** [comment]
At least in Germany, most "normal"/cheap dial-up lines have a forced reconnect all 24h -- if you don't want that (or even a static IP), you have to buy expensive business plans. We have the wonderful word "Zwangstrennung" for this... :(

**Comment by gdamjan** [comment]
@grawity  ADSL or even GPON services, T-Home macedonia and VIP.

I think they're doing that to encourage businesses to buy more expensive services

**Comment by JensErat** [comment]
> Maybe the wireguard option "persistent keepalive" will help, to resolve the problem?

Not really, since the IP addresses of the targets potentially change and Wireguard does not re-resolve them until now. This works somewhat fine and helps in case one end has a static IP, the dynamic side will keep open the connection and reconnect successfully if required. The static side will reliably update it's internal information when a client reconnects from another IP.

If both sides have dynamic IPs though (for example, laptop with mobile uplink on the road and a home connection with 24h-reconnect) both sides change their IPs with some bad timing and no side can reconnect any more.

Even worse: only one side is configured to connect (which would probably be the case in the mobile warrior scenario above): if the "server side" changes IPs, the connection will drop and not reconnect any more until manual intervention.

I run a wireguard VPN between two connections with 24h reset and it works out pretty stable _unless_ both connections reconnect pretty much at the same time -- at least it fails only very sporadically (all few weeks). Maybe some network operator maintenance or something like this until either networkd or wireguard get support for re-resolving hostnames. Looking into putting together some kind of "wireguard watch and reconnect" daemon though -- shouldn't be too hard, just disabling/enabling the interface is sufficient.

**Comment by pvcbe** [comment]
Hello, I made a temporary (shell) fix for systemd-networkd *.netdev files. 
https://github.com/pvcbe/systemd-wireguard-refresh

**Comment by ZLima12** [comment]
> Maybe the wireguard option "persistent keepalive" will help, to resolve the problem?

I had the same idea myself. You could have a persistent keepalive on both your client and server, so if either has a change of IP address, the other would be made aware when the keepalive gets sent. I may end up using this, but in the event that both devices' IP address change within the persistent keepalive window, it would be impossible to recover. Obviously this is extremely unlikely, but still feels like a bit of an oversight.

As for the discussion on whether this should be managed in userspace or not, a very large portion of Wireguard users would be positively affected by this option. As Wireguard itself is intentionally designed to be lightweight (to avoid the complexity of other VPN suites), I don't feel that this functionality belongs in Wireguard. That leaves either the system you're using to manage the Wireguard interface, or an external program that only refreshes DNS. The latter option seems a bit odd to me, as such a large number of users would have to install additional software for something so common.

Additionally, in the event of multiple Wireguard links, you may wish to only enable DNS refreshing for some. If you were using an external program for this, you would likely have to specify which interfaces to manage in a separate config file. It makes a lot more sense to me to specify this option for each interface in the netdev file.

**Comment by aburch** [comment]
This also happens when not using DNS names, but IP addresses for the WireGuard peer in `Endpoint=`.  It would be nice if WireGuard could keep track of a IPv4 and v6 peer address.
Re-resolving the DNS name would not address this use.

**Comment by lachesis** [comment]
The lack of this feature just borked all of my remote dial-home servers because my main box changed IP with a few hours downtime. Luckily I had a second connection that I could use to reprovision (really just restart networking). It's easy to add and would be useful - please add it.

**Comment by 1rabbit** [comment]
> The lack of this feature just borked all of my remote dial-home servers because my main box changed IP with a few hours downtime. Luckily I had a second connection that I could use to reprovision (really just restart networking). It's easy to add and would be useful - please add it.

I had to same exact issue today. Please add it to the windows and android clients, at least, where running scripts isn't a viable option, and feels like a band aid for something that could be easily baked in as an option.

**Comment by gdamjan** [comment]
> Please add it to the windows and android clients

you can't run systemd on android or windows

**Comment by bar-tolini** [comment]
I have the same problem with dynamic IPs in the picture with a different twist. 

WG is running on a machine in my LAN. I want to content to it from a mobile android device. 

My mobile device, while on the same LAN, users a local DNS to resolve the ddns address I use to connect to WG to a local IP (I can't do DNS hairpin).

**Comment by luochen1990** [comment]
> connection reestablishment on keepalive fail but without using DNS cache

I think that is a good idea,    this  solution will introduce no extra cost for people who has a static IP,   and make people who use dynamic IP feels much better, maybe we can add no option and just make this the default behavior.

For now,  I just add a cron tab line to run following command periodically as workaround:

```
wg set wg0 peer xtVrhteD7Txxxx= endpoint mydomain.com:51820
```

**Comment by bar-tolini** [comment]
As crazy as it sounds, an update to Android 12 fixed this issue for me. Apparently the OS does some magic to work around this problem now.

**Comment by WGH-** [comment]
On November 2, 2021 10:12:56 AM GMT+03:00, bar-tolini ***@***.***> wrote:
>As crazy as it sounds, an update to Android 12 fixed this issue for me. Apparently the OS does some magic to work around this problem now.
>

Android doesn't run systemd-networkd. Your issue is not this issue.

**Comment by 1rabbit** [comment]
> > Please add it to the windows and android clients
> 
> you can't run systemd on android or windows

Yes I am aware of that. I haven't referred to systemd at any given point so I'm unsure what you were assuming when typing this.

Anyway, I would like to know if there's anything we can do to help the situation on Android and windows devices regarding this problem.

**Comment by yuwata** [comment]
Hi, I've created PR #21562. If possible, please test the PR. Any comments and suggestions are welcome. Thank you.

**Comment by Majiir** [comment]
> However can we allow `networkctl reload/renew` to resolve dns for peer endpoint?

Some way to manually refresh a peer address with `networkctl` would be a great start.

Users (or distros) that want to periodically refresh could do so with a simple timer instead of resorting to hacky config file parsing and `wg set`.

See also https://github.com/systemd/systemd/issues/25547. It seems networkd does very little for reconfiguring Wireguard interfaces. The best known workaround for refreshing peers seems to be

```
ip link delete <interface> && networkctl reload
```

which will take down all peers in the process and maybe cause other problems.

**Comment by Tecol87** [comment]
> It's not entirely obvious that re-resolution is right for all WireGuard integrations
I also think it should be an optional feature.

> OpenVPN has an option for it, actually
Also OpenVPN will notice, when the connection drops (and can exit so the service can restart). I don't think wireguard can do this.

> Maybe the wireguard option "persistent keepalive" will help,
This option is intended to keep the firewall ports open. There is no option like disconnect and connect again if the handshakes fails.

> I don't feel that this functionality belongs in Wireguard
I agree. I think the service managing the interface should mange the interface.

**Comment by main--** [comment]
Here is the systemd equivalent of the reresolve-dns.sh script that ships with wireguard-tools, for use with a timer unit as described in [arch wiki](https://wiki.archlinux.org/title/WireGuard#Endpoint_with_changing_IP):

```
[Unit]
Description=Reresolve DNS of all WireGuard endpoints
Wants=network-online.target
After=network-online.target nss-lookup.target

[Service]
Type=oneshot
ExecStart=/bin/sh -c 'for if in $(networkctl list --json=pretty | jq -r \'.Interfaces[] | select(.Kind == "wireguard") | .Name\'); do networkctl cat "@$if:netdev" | awk -F \'\' \'/^PublicKey/ { pk = substr($0, index($0, "=")+1) } /^Endpoint/ { ep = substr($0, index($0, "=")+1) } /WireGuardPeer/ { if (pk) { print pk, ep } } END { print pk, ep }\' | while read pk ep; do wg set "$if" peer $pk endpoint $ep; done; done'
```

Requires jq, awk, and wg of course. The parsing is not very robust but should work with a sane netdev file.
